python/qa_BasicTest_XML.py

```diff
@@ -0,0 +1,259 @@
+import collections, datetime
+import os.path
+import xmltodict
+
+from gnuradio import gr, gr_unittest, blocks
+
+TestResultData    = collections.namedtuple('TestResultData', ['fromfile', 'passing', 'total', 'rate'])
+TestSerieSettings = collections.namedtuple('TestSerieSettings', ['data', 'times'])
+
+class qa_BasicTest_XML (gr_unittest.TestCase):
+
+    #
+    #   Listen on socket for data, append to list if any and return list of captured data.
+    #
+    def gatherFromSocket(self, amount):
+        total_data = []
+        data = ''
+
+        for i in range(amount):
+            try:
+                data = self.server.recv(4096)
+                if data:
+                    total_data.append(data)
+            except:
+                pass
+
+        return total_data
+
+    #
+    #   Compare captured data list with the expected data list.
+    #   Print received data, expected data and if they match.
+    #   Also give feedback about successrate in the decoding (percentage correctly captured and decoded).
+    #
+    def compareDataSets(self, total_data, expected_data, fromfile, cmpHDR):
+        global testResults
+        total_passing = 0
+
+        print '\nReceived:'
+
+        for idx, val in enumerate(expected_data):
+            data_str = '?'
+            passed = True
+
+            try:
+                data_str = " ".join("{:02x}".format(ord(n)) for n in total_data[idx])
+                self.assertEqual(val, data_str if cmpHDR else data_str[9:])
+            except:
+                passed = False
+
+            if passed:
+                total_passing += 1
+                print "{0: 3d} :: match:\n {1:s}".format(idx, val)
+            else:
+                print "{0: 3d} :: mismatch:\n {1:s}".format(idx, data_str)
+                print "should be:\n {0:s}".format(expected_data[idx])
+
+        results = TestResultData(fromfile, total_passing, len(expected_data), float(total_passing) / len(expected_data) * 100.0)
+        testResults[len(testResults) - 1].append(results)
+        print ("\nPassed rate: {0:d} out of {1:d}  ({2:.2f}%)\n"
+                .format(results[1], results[2], results[3]))
+
+    #
+    #   Connect a filesource with the given file and run the flowgraph.
+    #   File is acquired from the file_sink block and saved as .cfile (complex IQ data).
+    #
+    def runWithFileSourceAndData(self, infile, SpreadingFactor, cmpHDR):
+        print "Starting test from data in: {0:s}\n".format(infile)
+
+        if os.path.isfile(infile):
+            self.file_source.close()
+            self.file_source.open(infile, False)
+            self.sf = SpreadingFactor
+            self.lora_lora_receiver_0.set_sf(self.sf)
+            self.tb.run ()
+
+            total_data = self.gatherFromSocket(test_series[len(test_series) - 1][1])
+            self.compareDataSets(total_data,
+                                 test_series[len(test_series) - 1][0] * test_series[len(test_series) - 1][1],
```

Comment:
An entry in `test_series` constains the data to be checked and the amount of times it is expected.
Before a test is run, these are read from the xml file and appended to `test_series` so that the last element from `test_series` contains the data to compare to for the nexts series of tests.

`test_series` is later used to print to expected data for each series of test to the log file.

`infile` (not shown here) is a class variable in the new pull request.